### ISSUE TITLE
Feature proposal: "Dark mode" divergent colormaps

### ISSUE THREAD

**Issue by martinosorb** [matplotlib/matplotlib#18608]
Since version 2.0, matplotlib has done a good job encouraging users to use perceptually uniform colormaps, distinguishing between divergent and sequential colormaps, etc. In my job, I find divergent colormaps very useful; there is a good selection of them, but they all have a very light "zero" colour: typically white or yellow, except one which is light gray. They can be seen [here](https://matplotlib.org/3.3.0/_images/sphx_glr_colormaps_004.png). This is an example of what can be done:
![light](https://user-images.githubusercontent.com/11293950/94562053-23c50900-0265-11eb-87cb-9b66334de927.png)

With the explosion of "dark mode" in apps, websites and in the tastes of end-users, it would be very nice to have a divergent colormap that has a dark default colour.

### Proposed Solution

To add two or more colormaps which fulfill all the conditions of divergent colormaps, but have a dark colour as a middle colour. At least one of these should use pure black.

### Additional context and prior art

Currently, the closest approximation is `plt.cmap.twilight`, however, it does not look sufficiently symmetric to me. Moreover, it's cyclic, so it ends with the same colour on both sides, which is undesirable. This is how it looks like: 
![dark](https://user-images.githubusercontent.com/11293950/94562105-33dce880-0265-11eb-8f95-e26dc8c6ff48.png)

It's quite close to what I'm suggesting, except for the details I mentioned above (and I would add one or two more, including at least one with a black background).

**Comment by mwaskom** [comment]
I don't think matplotlib has any built-in perceptually uniform divergent maps. The ones you linked to are mostly drawn ColorBrewer and have pretty good properties, but they are not perceptually uniform in the same color space as viridis et al.

There are several existing implementations, mostly with relatively low contrast (i.e. the extremes are medium-luminance saturated colors) in [colorcet](https://colorcet.holoviz.org/).

seaborn also has a high-contrast diverging map with a dark center ("icefire"): http://seaborn.pydata.org/tutorial/color_palettes.html#diverging-color-palettes

**Comment by jklymak** [comment]
I agree that we don't have a dark-centred diverging colormap, and I think that would be a reasonable addition. 

I think its fair to say the consensus is that matplotlib should be conservative about adding new colormaps.   It stops our colormap palette from becoming unmanageable, and saves PRs where pet colormaps are proposed, folks argue about which shade of chartreuse should be used, or whether the Smithetal02 definition of colormap uniformity is the right model or Jonesetal09.  

The best way to get a colormap in is to argue that it is unique from the offered ones.  i.e. `twilight` was properly cyclic, and that was deemed unique enough.  These days it helps to argue perceptual uniformity, though I personally don't think that should be a deal-killer.   It really helps to show that it has already been used in publications somewhere, or that it is very similar to published data.

**Comment by martinosorb** [comment]
The consensus seems reasonable in not wanting to add too many new colormaps. However I do think that this is, as you say, something unique and substantially different from all the existing ones, and not just a variation on shades.

I'm not an expert in colour and I'm not familiar with the literature -- I'm suggesting this in my role as an end user of matplotlib. However, I think a choice similar to Seaborn's `icefire`, suggested by @mwaskom above, would be great. It's divergent, perceptually uniform, and I think likely colorblind-safe.

**Comment by ingwag** [comment]
Hi there are several packages available with perceptually uniform colourmaps. I'm aware of CMasher
https://joss.theoj.org/papers/10.21105/joss.02004
and Fabio Crameri's scientific colormaps
https://www.nature.com/articles/s41467-020-19160-7
https://www.fabiocrameri.ch/colourmaps/

I think adding diverging colormaps with dark central colors would be a great addition to matplotlib, it would certainly be useful for some outreach activities (at least the one that I'm involved in)

**Comment by hlongmore** [comment]
_(Note: most of this comment is to help anyone looking for a workaround until there are other colormaps for dark mode, but also to resurface the issue and provide additional motivation)_

Looking at https://matplotlib.org/stable/users/explain/colors/colormaps.html and trying to make a dark mode heat map plot using `imshow` that works just as well on the screen in a Jupyter notebook as when saved to file, I'm noticing that:

- All of the Perceptually Uniform Sequential colormaps have some shade of yellow-ish at one endpoint, making it an awful choice for white text if you have data at that extreme
- All of the Sequential colormaps have white at one endpoint, making it a bad choice to set the text to white if you have data at that extreme
- Almost all of the Sequential (2) have the same problem, some with white, some with yellow; the exceptions of winter, cool, and copper also have an endpoint that doesn't look good with white text, though they are better than yellow or white endpionts
- As already mentioned, the Divergent colormaps have a very light "zero color", usually white, maybe yellow, so again, not a good choice for white text
- The rest of the colormaps, while not being a good choice for a heat map, also have white or very light colors, and/or yellow
- A quick perusal of the links in ingwag's comment above shows they all basically have the same issues: either a light or a white or a yellow endpoint, or a midpoint of the same.

All the above notwithstanding, white text covers the most ground. My current workaround is to abuse the vmin/vmax.
With `winter` and `Purples_r` I can only use about 50%-60% of the colormap range, but with `viridis` and `gist_heat` I can use about 80% of the range (maybe 85%, but that's pushing it); and how well that works depends on how much variance there is in my data. Here's the gist of my workaround:

```python
    # data: pandas.DataFrame
    colormap = 'gist_heat'
    plt.style.use('dark_background')

    # Determine vmin / vmax to make the colormap work better (no white text on yellow / light green / white)
    max_value = max(data.max())
    min_value = min(data.min())
    value_range = max_value - min_value
    full_range = value_range / 0.8
    difference = full_range - value_range
    desired_vmin = min_value
    desired_vmax = max_value + difference

    fig, ax = plt.subplots()
    im = ax.imshow(
        data,
        cmap=colormap,
        vmin=desired_vmin,
        vmax=desired_vmax,
    )
```

I really like the `Purples_r` in dark mode, and if I use a `PowerNorm` with `gamma=1.5` I can use closer to 75% of the range with pretty decent results, but still pushing it on my max values, and if I have data clustered toward vmin, there's not a lot of color variation for the lower values. Using a similar technique with `viridis`  also works to avoid the yellow endpoint and use slightly more of the range.

Another option (having just discovered `AsinhNorm` while tweaking things) would be to use `AsinhNorm`, something like:
```python
    # data: pandas.DataFrame
    colormap = 'gist_heat'
    plt.style.use('dark_background')
    
    # Determine vmin / vmax to make the colormap work better (no white text on yellow / light green / white)
    max_value = max(data.max())
    min_value = min(data.min())
    std_dev = data.stack().std()
    value_range = max_value - min_value
    full_range = value_range / 0.7
    difference = full_range - value_range
    desired_vmin = min_value
    desired_vmax = max_value + difference
    desired_linear_width = std_dev * ((value_range * 0.75) / std_dev)
    # print(f'{desired_vmin=}, {desired_vmax=}, {desired_linear_width=}, {std_dev=}')

    fig, ax = plt.subplots()
    im = ax.imshow(
        data,
        cmap=colormap,
        norm=colors.AsinhNorm(
            linear_width=desired_linear_width,
            vmin=desired_vmin,
            vmax=desired_vmax,
            clip=False,
        )
    )
```
So far, this works great with `gist_heat` and `viridis`, and pretty good for `winter`. With `Purples` and `Purples_r` if I reduce the range to 35%-40%, and increase the `value_range` multiplier for the `desired_linear_width` to `0.95`,  the white text is more visible, and the color variance in my low-value clusters is better (and almost good enough to use purples instead of `viridis` or `gist_heat`). Those same settings also wok oaky with `viridis` and `gist_heat`, but not as well as the 0.7 / 0.75 in the code example above (0.7 / 0.9 also works well).

If there were a way to adjust the colormap by saying "Give me this colormap, but ignore the top X%, and map my data to the rest of the colormap" that would be cool. Alternatively, new colormaps that effectively did the same, so that the troublesome white or yellow endpoints could be removed, would also work.

**Comment by martinosorb** [comment]
I looked at @ingwag's suggestion of Fabio Crameri's work (@crameri), which seems to be scientifically based and well thought of. First, I note that in [his paper](https://doi.org/10.1038/s41467-020-19160-7)'s "Guideline for choosing the right scientiﬁc colour map" (figure 6), we are clearly missing the dark-coloured-centre colourmap.

These are his colourmaps:
![image](https://i0.wp.com/s-ink.org/wp-content/uploads/ScientificColourMaps8-FabioCrameri.png)

I suggest we need a couple:
- one that has black as a center colour, and one that doesn't
- one that has blue and red as lateral colours, and one with less symbolic colours
So having _vanimo_ and _managua_ or something like that could be a choice.

Note that  he releases some matplotlib code ready made with those colormaps, under MIT license.  Would it be feasible to add it as-is to matplotlib, carrying the MIT license over? Otherwise, we should make our own similar colormaps.

**Comment by Crameri** [comment]
If you are interested, I would be very happy to have some (or all) of the Scientific colour maps implemented in Matplotlib and made directly available to your users as long as (i) a reference to the repository and the specific version number of the included Scientific colour maps (see https://zenodo.org/records/5501399) are stated somewhere, and (ii) the names of the colour maps are maintained.

**Comment by martinosorb** [comment]
This is up to the maintainers then. I could _try_ (I'm not that familiar with mpl source) to submit a PR, if the maintainers are ok with this and if they can suggest the number of such colourmaps that could be accommodated.

**Comment by jklymak** [comment]
I think a subset of these that are distinct from those in Matplotlib would be welcome.  

> (i) a reference to the repository and the specific version number of the included Scientific colour maps

This seems fine, but where do you just mean in the source code?  Because we don't really have documentation for individual colormaps, though linking out to your collection would be great in the narrative docs (eg Choosing a Colormap)
